### PR TITLE
JAX TD3 prototype

### LINKED ISSUE #218: JAX Integration with CleanRL
## Problem Description
Given the incredible performance of the DDPG + JAX prototype (https://github.com/vwxyzjn/cleanrl/pull/187), it's worth prototyping JAX with other algorithms as well! This issue tracks the overall progress of integrating JAX with CleanRL.

### Useful resources

* (a working JAX + DDPG example as a reference implementation) https://github.com/vwxyzjn/cleanrl/pull/187
    * CleanRL's DDPG docs: https://docs.cleanrl.dev/rl-algorithms/ddpg/
* (a working JAX + PPO example as a reference implementation) #217 
    * CleanRL's PPO docs: https://docs.cleanrl.dev/rl-algorithms/ppo/

Common gotchas and errors:
- https://github.com/google/jax/issues/2697#issuecomment-613181201
- https://github.com/deepmind/optax/issues/366#issuecomment-1166632881

### Useful pattern when extending

In CleanRL a filediff is incredibly helpful. For example, if I want to learn how TD3 is different from DDPG, I could do 
- open VS code and select `ddpg_continuous_action.py` and `td3_continuous_action.py`
- right-click and left-click "compare selected"
- the following file diff window shows up

image

### Contribution process

There is a contribution checklist to help streamline the contribution process. For each new contribution, we'd need to add documentation, tests, run benchmark experiments, etc. See https://github.com/vwxyzjn/cleanrl/pull/186 as an example.

### Tracked issues
- [x] #216 
- [ ] #217 
- [x] #219
- [x] #220
- [ ] #221

### PR DESCRIPTION
**Pull request by joaogui1**
## Description
Closes #218 
Initiali implementation, needs testing

## Types of changes

- [x] New feature
- [x] New algorithm

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/vwxyzjn/cleanrl/blob/master/CONTRIBUTING.md) guide (**required**).
- [x] I have ensured `pre-commit run --all-files` passes (**required**).
- [x] I have updated the documentation and previewed the changes via `mkdocs serve`.
- [x] I have updated the tests accordingly (if applicable).

If you are adding new algorithms or your change could result in performance difference, you may need to (re-)run tracked experiments. See https://github.com/vwxyzjn/cleanrl/pull/137 as an example PR. 
- [x] I have contacted @vwxyzjn to obtain access to the [openrlbenchmark W&B team](https://wandb.ai/openrlbenchmark) (**required**).
- [x] I have tracked applicable experiments in [openrlbenchmark/cleanrl](https://wandb.ai/openrlbenchmark/cleanrl) with `--capture-video` flag toggled on (**required**).
- [x] I have added additional documentation and previewed the changes via `mkdocs serve`.
    - [x] I have explained note-worthy implementation details.
    - [x] I have explained the logged metrics.
    - [x] I have added links to the original paper and related papers (if applicable).
    - [x] I have added links to the PR related to the algorithm.
    - [x] I have created a table comparing my results against those from reputable sources (i.e., the original paper or other reference implementation).
    - [x] I have added the learning curves (in PNG format with `width=500` and `height=300`).
    - [x] I have added links to the tracked experiments.
    - [x] I have updated the overview sections at the [docs](https://docs.cleanrl.dev/rl-algorithms/overview/) and the [repo](https://github.com/vwxyzjn/cleanrl#overview)
- [x] I have updated the tests accordingly (if applicable).

### CODE REVIEW

File: cleanrl/td3_continuous_action_jax.py [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+        # TRY NOT TO MODIFY: CRUCIAL step easy to overlook
+        obs = next_obs
+
+        # ALGO LOGIC: training.
+        if global_step > args.learning_starts:
+            data = rb.sample(args.batch_size)
+            # TODO Maybe generate a lot of random keys right in the beginning
+            # also check https://jax.readthedocs.io/en/latest/jax.random.html
+            key, noise_key = jax.random.split(key, 2)
+            clipped_noise = jnp.clip((jax.random.normal(
```
Comment (vwxyzjn) [inline cleanrl/td3_continuous_action_jax.py:287]: `clipped_noise` is not actually used. Also, maybe generating it with numpy is a little bit faster? With `jnp` we would need to jit function probably. Would you mind doing a speed test like `%timeit`?

File: cleanrl/td3_continuous_action_jax.py [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+        actor_state: TrainState,
+        qf1_state: TrainState,
+        qf2_state: TrainState,
+        observations: np.ndarray,
+        actions: np.ndarray,
+        next_observations: np.ndarray,
+        rewards: np.ndarray,
+        dones: np.ndarray,
+    ):
+        next_state_actions = (actor.apply(actor_state.target_params, next_observations)).clip(-1, 1)  # TODO: proper clip
```
Comment (vwxyzjn) [inline cleanrl/td3_continuous_action_jax.py:202]: Yeah so if we want to use `jnp` to generate the clipped_noise, we should probably do it here.

File: cleanrl/td3_continuous_action_jax.py [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+        tx=optax.adam(learning_rate=args.learning_rate),
+    )
+    qf1 = QNetwork()
+    qf1_state = TrainState.create(
+        apply_fn=qf1.apply,
+        params=qf1.init(qf1_key, obs, envs.action_space.sample()),
+        target_params=qf1.init(qf1_key, obs, envs.action_space.sample()),
+        tx=optax.adam(learning_rate=args.learning_rate),
+    )
+    qf2 = QNetwork()
```
Comment (vwxyzjn) [inline cleanrl/td3_continuous_action_jax.py:180]: also do we need a `qf2`? We could maybe do 

```python
qf = QNetwork()
qf1_state = TrainState.create(
    apply_fn=qf.apply,
    params=qf.init(qf_key, obs, envs.action_space.sample()),
    target_params=qf.init(qf_key, obs, envs.action_space.sample()),
    tx=optax.adam(learning_rate=args.learning_rate),
)
qf2_state = TrainState.create(
    apply_fn=qf.apply,
    params=qf.init(qf_key, obs, envs.action_space.sample()),
    target_params=qf.init(qf_key, obs, envs.action_space.sample()),
    tx=optax.adam(learning_rate=args.learning_rate),
)
actor.apply = jax.jit(actor.apply)
qf.apply = jax.jit(qf.apply)
```

File: cleanrl/td3_continuous_action_jax.py [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+        def mse_loss(params, qf):
+            qf_a_values = qf.apply(params, observations, actions).squeeze()
+            return ((qf_a_values - next_q_value) ** 2).mean(), qf_a_values.mean()
+
+        (qf1_loss_value, qf1_a_values), grads1 = jax.value_and_grad(mse_loss,
+                                                    has_aux=True)(qf1_state.params, qf1)
+        (qf2_loss_value, qf2_a_values), grads2 = jax.value_and_grad(mse_loss,
+                                                    has_aux=True)(qf2_state.params, qf2)
+        qf1_state = qf1_state.apply_gradients(grads=grads1)
+        qf2_state = qf2_state.apply_gradients(grads=grads2)
```
Comment (vwxyzjn) [inline cleanrl/td3_continuous_action_jax.py:217]: We are doing grad passes twice. Would it be faster to have them share the same optimizer as done in here?

https://github.com/vwxyzjn/cleanrl/blob/399f9a36e9d48032102b68078b1ca1cfaa940ba6/cleanrl/ppo_continuous_action_envpool_jax.py#L213-L217

**Review by vwxyzjn — COMMENTED** [review]
Thanks for the PR. Looks great!

File: cleanrl/td3_continuous_action_jax.py (hunk 3 above)
Comment (joaogui1) [reply cleanrl/td3_continuous_action_jax.py]: good idea, just changed it!

File: cleanrl/td3_continuous_action_jax.py (hunk 1 above)
Comment (joaogui1) [reply cleanrl/td3_continuous_action_jax.py:286]: Added it and fixed clipping, will test what's the most efficient way to generate noise soonish

**Review by vwxyzjn — COMMENTED** [review]
Hey, I added some changes. Overall it looks pretty good, but I don't know why the experiments do not work yet...

File: cleanrl/td3_continuous_action_jax.py [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+    ):
+        def actor_loss(params):
+            return -qf.apply(qf1_state.params, observations, actor.apply(params, observations)).mean()
+
+        actor_loss_value, grads = jax.value_and_grad(actor_loss)(actor_state.params)
+        actor_state = actor_state.apply_gradients(grads=grads)
+        actor_state = actor_state.replace(
+            target_params=optax.incremental_update(actor_state.params, actor_state.target_params, args.tau)
+        )
+        # TODO maybe split this in 2 functions?
```
[the remark this answers is not in the document]
Comment (joaogui1) [reply cleanrl/td3_continuous_action_jax.py:236]: I think it was related to splitting into two calls to update, one for qf1 and another for qf2, but it's def not worth it

File: cleanrl/td3_continuous_action_jax.py [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+        if global_step > args.learning_starts:
+            data = rb.sample(args.batch_size)
+            # TODO Maybe generate a lot of random keys right in the beginning
+            # also check https://jax.readthedocs.io/en/latest/jax.random.html
+            key, noise_key = jax.random.split(key, 2)
+            clipped_noise = jnp.clip(
+                (jax.random.normal(noise_key, actions[0].shape) * args.policy_noise),
+                -args.noise_clip,
+                args.noise_clip,
+            )
```
[the remark this answers is not in the document]
Comment (joaogui1) [reply cleanrl/td3_continuous_action_jax.py]: Right, done!

File: cleanrl/td3_continuous_action_jax.py [hunk 7]
```diff
@@ -248,7 +250,7 @@ def actor_loss(params):
         qf2_state = qf2_state.replace(
             target_params=optax.incremental_update(qf2_state.params, qf2_state.target_params, args.tau)
         )
-        return actor_state, qf1_state, actor_loss_value
+        return actor_state, qf1_state, actor_loss_value, grad_norm
```
Comment (vwxyzjn) [inline cleanrl/td3_continuous_action_jax.py]: I suppose the `grad_norm` is for debugging purposes? Please remove it in the final version.

**Review by dosssman — APPROVED** [review]
Great addition.

Looking good on my side

**Comment by vwxyzjn** [comment]
@joaogui1 could you take a final look at https://cleanrl-git-fork-joaogui1-master-vwxyzjn.vercel.app/rl-algorithms/td3/#td3_continuous_action_jaxpy to see if there is anything missing?

**Comment by joaogui1** [comment]
LGTM!

**Review by vwxyzjn — APPROVED** [review]
Thanks so much for this contribution!

**Pull request merged**